File: graphql-api/src/main/java/io/prizy/publicapi/graphql/scalars/LocalDateScalar.java
package io.prizy.publicapi.graphql.scalars;

import graphql.language.StringValue;
import graphql.schema.Coercing;
import graphql.schema.CoercingParseLiteralException;
import graphql.schema.CoercingParseValueException;
import graphql.schema.CoercingSerializeException;
import graphql.schema.GraphQLScalarType;
import lombok.experimental.UtilityClass;

import java.time.LocalDate;

/**
 * @author Nidhal Dogga
 * @created 6/24/2022 9:21 PM
 */


@UtilityClass
public class LocalDateScalar {

  public GraphQLScalarType SCALAR = GraphQLScalarType.newScalar()
    .name("Date")
    .description("Date")
    .coercing(new InstantCoercing())
    .build();

  private class InstantCoercing implements Coercing<LocalDate, String> {

    @Override
    public String serialize(Object dataFetcherResult) throws CoercingSerializeException {
      if (dataFetcherResult instanceof LocalDate instant) {
        return instant.toString();
      }
      throw new CoercingSerializeException("Cannot serialize: " + dataFetcherResult);
    }

    @Override
    public LocalDate parseValue(Object input) throws CoercingParseValueException {
      return LocalDate.parse(input.toString());
    }

    @Override
    public LocalDate parseLiteral(Object input) throws CoercingParseLiteralException {
      if (input instanceof StringValue stringValue) {
        return LocalDate.parse(stringValue.getValue());
      }
      throw new CoercingParseLiteralException("Cannot parse value: " + input);
    }
  }

}

File: graphql-core/src/main/kotlin/io/prizy/graphql/scalars/LocalDateScalar.kt
package io.prizy.graphql.scalars

import graphql.language.StringValue
import graphql.schema.Coercing
import graphql.schema.CoercingParseLiteralException
import graphql.schema.CoercingSerializeException
import graphql.schema.GraphQLScalarType
import java.time.LocalDate

object LocalDateScalar {

  val graphqlScalarType: GraphQLScalarType = GraphQLScalarType.newScalar()
    .name("Date")
    .description("Date")
    .coercing(InstantCoercing)
    .build()

  object InstantCoercing : Coercing<LocalDate, String> {

    override fun serialize(dataFetcherResult: Any): String {
      when (dataFetcherResult) {
        is LocalDate -> return dataFetcherResult.toString()
        else -> throw CoercingSerializeException("Cannot serialize $dataFetcherResult")
      }
    }

    override fun parseValue(input: Any): LocalDate {
      return LocalDate.parse(input.toString())
    }

    override fun parseLiteral(input: Any): LocalDate {
      when (input) {
        is StringValue -> return LocalDate.parse(input.value)
        else -> throw CoercingParseLiteralException("Cannot parse value: $input")
      }
    }
  }
}